Add/sites thumbnails

vykes-mac opened the pull request:
#### Proposed Changes

* Add site thumbnail component in sites dashboard

#### Testing Instructions

- Go to `/sites-dashboard`
- Verify sites thumbnail shows mshot image for launched sites
- Verify sites thumbnail shows site icon for private/coming soon sites when icon is present
- Verify sites thumbnail shows first letter of site domain when site icon is not present
- Verify background color is selected from site icon for private/coming soon sites

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

***Limitation**
Currently the hook that gets the primary color from the site icon encounters CORS issues when fetching image from atomic and some other sites (see console logs). Presently those sites retain a default background and colors are only pick from sites with access to the site icon.

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have you sent any new strings for translation (PCYsg-1vr-p2) ASAP?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

**Note**: background color now is a randomly generated color, that will be fixed in another PR

Related to https://github.com/Automattic/wp-calypso/issues/65171 and https://github.com/Automattic/wp-calypso/issues/65942

vykes-mac committed:
add site thumbnail to sites dashboard

sejas commented on client/sites-dashboard/components/sites-table-row.tsx:
```diff
@@ -139,7 +140,17 @@ export default function SitesTableRow( { site }: SiteTableRowProps ) {
+									<SiteThumbnail
+										mShotsUrl={ setmShotsUrl ? site.URL : undefined }
+										alt={ site.name }
+										backgroundColor={ '#' + Math.floor( Math.random() * 16777215 ).toString( 16 ) }
```
I suggest using a deterministic way to generate a color using the `site.ID`.

Something like this:
```js
'#' + Math.floor( site.ID * 999999 )
    .toString( 16 )
    .substring( 0, 6 )
```

Using a random color could confuse the user, and It would change in each render.

p-jackson commented:
This looks really cool!

I've got a lot of sites with white backgrounds so the thumbnail just sort of bleeds into the rest of the page.
<img width="489" alt="CleanShot 2022-08-02 at 14 03 35@2x" src="https://user-images.githubusercontent.com/1500769/182276017-48f76d59-5f9d-4a5a-9c71-4462d14ab560.png">

I put a very light border on with
```css
  border: 1px solid rgb(238, 238, 238);
  box-sizing: border-box;
```
and got this

<img width="452" alt="CleanShot 2022-08-02 at 14 04 35@2x" src="https://user-images.githubusercontent.com/1500769/182276100-64726216-4845-483c-8c6c-ce507a4077a5.png">


I matched the border with the separator between table rows, but maybe it should go even darker. What do you think?

vykes-mac commented:
@p-jackson yeah that definitely helps the thumbnail to pop more, I'm going to apply it to those thumbnails with colored background to see how it looks on those.

vykes-mac committed:
add react hook to fetch primary color from a given image

vykes-mac committed:
get site icon dominant color as background for non public sites

vykes-mac commented on package.json:
```diff
@@ -188,6 +188,7 @@
 		"config": "^3.3.6",
 		"debug": "^4.3.3",
 		"enhanced-resolve": "5.9.3",
+		"fast-average-color": "^6.4.2",
```
I'm not certain if this package should be added here or should be added `packages/components/package.json`

p-jackson replied:
It should be added in `packages/components/package.json`
https://github.com/Automattic/wp-calypso/blob/f7d45d6dc43d703a2f39ec8efd9a44097cf6a25f/docs/dependency-management.md#L28

vykes-mac committed:
update version of fast-average-color to accept CORS value

vykes-mac committed:
seperate processing an existing image from a provided url

vykes-mac committed:
use Image component in order to fetch private atomic site icon

vykes-mac commented:
In regards to the CORS limitation mentioned in the description, I think it only throws that error on local. Can someone confirm that thought?

sejas commented on packages/components/src/site-thumbnail/use-dominant-color.tsx:
```diff
@@ -0,0 +1,42 @@
+import FastAverageColor, { FastAverageColorResult } from 'fast-average-color';
+import { useEffect, useState } from 'react';
+
+export default function useDominantColor(
+	url?: string,
+	imageRef?: React.RefObject< HTMLImageElement >
+) {
+	const [ color, setColor ] = useState< FastAverageColorResult >();
+
+	useEffect( () => {
+		const fac = new FastAverageColor();
+		//get the dominant color of a loaded image
+		if ( imageRef && imageRef.current ) {
+			const color = fac.getColor( imageRef.current, {
+				ignoredColor: [
+					[ 255, 255, 255, 255 ], // white
+					[ 0, 0, 0, 255 ], // black
+				],
+				width: 50,
+				height: 50,
+			} );
+			setColor( color );
+		}
+
+		//get the dominant color of image from url
+		if ( url && url.length > 0 ) {
+			fac
+				.getColorAsync( url, {
+					ignoredColor: [
```
I think accepting two optional parameters can be confusing.
I suggest accepting only one `image: string | React.RefObject< HTMLImageElement >` and make a `if else if` based on the param type and its value.
Reading the current code seems that two parameters can be used simultaneously.

sejas replied:
Here is my PR of my experiment: https://github.com/Automattic/wp-calypso/pull/66261/files

vykes-mac replied:
> I've tried passing the image ref instead of the URL, but I'm getting the same CORS error.

The image ref option is really for local images that are not loaded from a server, I should probably remove that aspect for now. 
Even though CORS error is encountered the default background color is still used, However doesn't seem it's possible to prevent those issues from logging in the console.


> I think accepting two optional parameters can be confusing.
> I suggest accepting only one image: string | React.RefObject< HTMLImageElement > and make a if else if based on the param type and its value.
> Reading the current code seems that two parameters can be used simultaneously.

Yeah that's true. I don't think we really need the image ref at the moment, I'm only gonna use the url for now and can extend later to accept an image reference if needs be.

sejas reviewed (commented):
@vykes-mac 
About CORS, here is what I've found:

I'm getting the CORS error also on calypso live.
<img width="1276" alt="Screenshot 2022-08-04 at 13 46 27" src="https://user-images.githubusercontent.com/779993/182850463-d0297deb-79ac-4599-b75a-5fba2a98a8cd.png">

I've tried passing the image ref instead of the URL, but I'm getting the same CORS error.

The browser can display the image but doesn't allow it to read its content programmatically.

The error is launched when the library calls to:
```js
ctx.drawImage(image, 0, 0) // the image is rendered ok inside the canvas
ctx.getImageData(0, 0, width, height) // CORS error is triggered
```

The exact error is: 

`The canvas has been tainted by cross-origin data`

I've tried adding the `crossOrigin = 'anonymous'` to the image element and to the library options, but it didn't fix the error.

FastAverageColor returns a black color `#000000` when the CORS error occurs.



We could add the header `Access-Control-Allow-Origin "*"` in the backend when it returns the site logo. That would fix our problem, but It seems overkill.

I do not see an easy solution to extract the main color from the logo.

vykes-mac committed:
fetch image color from an image url

vykes-mac committed:
merge trunk into add/sites-thumbnails

p-jackson commented on packages/components/src/site-thumbnail/use-dominant-color.tsx:
```diff
@@ -0,0 +1,27 @@
+				} )
+				.then( ( color ) => setColor( color ) )
+				.catch( () => {
+					return;
```
I'm wondering if it's possible to have any errors that happen here get reported in Sentry p4TIVU-a8b-p2
It doesn't look like there's an explicit `logToSentry` function in the codebase so maybe it just automatically logs uncaught errors?

@fullofcaffeine what's the best approach for marshalling a promise rejection like this onto the main "thread" so that it gets handled by Sentry (if that's even the way to do it). Is it just to `await` the promise?

p-jackson replied:
This isn't a blocker

vykes-mac replied:
there's a `captureException` from `'calypso/lib/sentry'` but it doesn't seem it should be used in `/packages/ `folder

p-jackson replied:
Nice find, yes this seems perfect I don't see why this couldn't be moved to `@automattic/sentry` tbh, but let's not propose that here.

p-jackson commented on packages/components/src/site-thumbnail/style.scss:
```diff
@@ -5,6 +5,12 @@
+  	-o-transition: background-color 1s ease-out;
+  	transition: background-color 1s ease-out;
+	border: 1px solid rgb( 238, 238, 238 );
+  	box-sizing: border-box;
```
Just a little white space thing here

```suggestion
	-webkit-transition: background-color 1s ease-out;
	-moz-transition: background-color 1s ease-out;
	-o-transition: background-color 1s ease-out;
	transition: background-color 1s ease-out;
	border: 1px solid rgb( 238, 238, 238 );
	box-sizing: border-box;
```

p-jackson commented on packages/components/src/site-thumbnail/use-dominant-color.tsx:
```diff
@@ -0,0 +1,27 @@
+	const [ color, setColor ] = useState< FastAverageColorResult >();
+
+	useEffect( () => {
+		const fac = new FastAverageColor();
```
[The `FastAverageColor` instance has a `.destroy()` method](https://github.com/fast-average-color/fast-average-color/blob/8992453cd92f6de723022395499650e7e41bc3f4/src/index.ts#L217) that needs to be called to clean up the internal canvas.

[`useEffect` has a way to run cleanup](https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect) which is to return a cleanup function. So it'd be something like this:

```ts
useEffect( () => {
	const fac = new FastAverageColor();

	// ... do the stuff that calls `setColor()`

	return () => {
		fac.destroy();
	};
} );
```

p-jackson commented on client/sites-dashboard/components/sites-table-row.tsx:
```diff
@@ -142,7 +143,22 @@ export default function SitesTableRow( { site }: SiteTableRowProps ) {
+										className={ undefined }
+									/>
+								) : (
+									<h1 style={ { fontSize: 'xx-large' } }>{ site.name.charAt( 0 ) }</h1>
```
`<h1>` isn't an accessible choice because a screen reader is going to announce all of these capital letters as being the main heading for the page.

Since the text is really being used as an icon I wonder if setting the ARIA role to `img` makes sense.

```suggestion
									<div role="img" style={ { fontSize: 'xx-large' } }>{ site.name.charAt( 0 ) }</div>
```

Since we're not modifying the style at runtime it's better to set the styling with CSS, so using the ``styled.div` ... `;`` syntax.

vykes-mac replied:
I think this make sense and set the aria-label to site icon or site logo?

p-jackson replied:
I think this already covered by the `<a>` element having the "Visit Dashboard" text. This icon isn't visually important, it's the link text that communicates what the link is for.

In saying that I think it's supposed to be [better to have the link text inside the `<a>`, not as a `title` attribute](https://a11y-101.com/development/icons-and-links). But that's unrelated to this PR.

p-jackson commented on client/sites-dashboard/components/sites-table-row.tsx:
```diff
@@ -142,7 +143,22 @@ export default function SitesTableRow( { site }: SiteTableRowProps ) {
+										src={ site.icon.img }
+										alt={ site.name }
+										style={ { height: '50px', width: '50px' } }
+										className={ undefined }
```
It's a bit strange having to pass `undefined` here. I'm guessing it's because you got a TypeScript error?

`<Image>` is implemented in JS and TypeScript is doing its best to infer the prop types, but it's doing a bad job and gotten it wrong. You can fix this by adding this esoteric jsdoc to the `Image` function definition in `image/index.js`

```js
/**
 * @param {import('react').ImgHTMLAttributes} props
 */
export default function Image( { alt, className, src, ...restProps } ) {
```

This says that the `<Image>` component takes all the same prop types as the `<img>` element.

This fixes things up so that we can leave `className` blank (and it should improve the typing for everyone else using `<Image>` too!

vykes-mac replied:
yeah I found it strange but typescript kept screaming about it.

p-jackson reviewed (changes requested):
It's looking good 👍 There are a few things to fix up, but once that's done I think we can merge it as is (even if some of the background colours have CORS errors) and we can tidy them up in other PRs.

p-jackson commented:
The `fast-average-color` package seems reasonably well maintained
https://snyk.io/advisor/npm-package/fast-average-color
https://isitmaintained.com/project/fast-average-color/fast-average-color

vykes-mac committed:
clean up canvas after image has been processed

vykes-mac committed:
clean up styling and typescript error for Image component

vykes-mac committed:
move dependency to the correct package.json file

vykes-mac committed:
fix css

zaguiini commented on client/components/image/index.js:
```diff
@@ -5,6 +5,9 @@ const hideImageOnError = ( event ) => {
 	event.target.style.display = 'none';
 };
 
+/**
+ * @param {import('react').ImgHTMLAttributes} props
+ */
```
I wonder if we could turn this into a TS file instead of adding pseudo-typing through JSDocs :)

vykes-mac replied:
would it affect imports in other places? Don't want that to caused a bunch of file changes in this PR

zaguiini replied:
No! At least it shouldn't :)

zaguiini commented on client/sites-dashboard/components/sites-table-row.tsx:
```diff
@@ -98,6 +98,10 @@ const ListTileSubtitle = styled.div`
 
+const NoIcon = styled.div`
+	font-size: xx-large;
+`;
```
Let's prefer CSS objects instead of string soups:

```suggestion
const NoIcon = styled.div({
	fontSize: 'xx-large'
})
```

CSS objects give us typing and syntax highlighting out of the box. It's much more TS friendly too 👍

zaguiini commented on client/sites-dashboard/components/sites-table-row.tsx:
```diff
@@ -129,6 +133,7 @@ export default function SitesTableRow( { site }: SiteTableRowProps ) {
 	} else if ( site.is_private ) {
 		siteStatusLabel = __( 'Private' );
 	}
+	const setmShotsUrl = ! isComingSoon && ! site.is_private;
```
How about using a more descriptive name here?

```suggestion
	const shouldUseScreenshot = ! isComingSoon && ! site.is_private;
```

vykes-mac replied:
yeah that make sense

zaguiini commented on client/sites-dashboard/components/sites-table-row.tsx:
```diff
@@ -129,6 +133,7 @@ export default function SitesTableRow( { site }: SiteTableRowProps ) {
 	} else if ( site.is_private ) {
 		siteStatusLabel = __( 'Private' );
 	}
+	const setmShotsUrl = ! isComingSoon && ! site.is_private;
```
We're starting to have a lot of `! isComingSoon && ! site.is_private` used throughout the SMD module and really it's just a way of checking if a site is public. We can do this later in a follow-up, but I'd be really interested in an abstraction that hides and turns this condition clearer within a single place (that would help in maintenance).

vykes-mac replied:
yeah we might have to look at that in another PR, would be nice it was already apart of the site object to tell when the site is already launched/public.

zaguiini commented on packages/components/src/site-thumbnail/style.scss:
```diff
@@ -5,6 +5,12 @@
 	justify-content: center;
 	position: relative;
 	overflow: hidden;
+	-webkit-transition: background-color 1s ease-out;
+	-moz-transition: background-color 1s ease-out;
+	-o-transition: background-color 1s ease-out;
```
Do we need these prefixed versions? I think the CSS post-processor adds them for us.

vykes-mac replied:
Yeah you are correct, I think autoprefixer adds them.

zaguiini commented on packages/components/src/site-thumbnail/index.tsx:
```diff
@@ -24,11 +26,16 @@ export const SiteThumbnail = ( {
 
+	const primary = useDominantColor( bgColorImgUrl );
+
+	backgroundColor = primary?.hexa || backgroundColor;
```
Let's try not to reassign the props object. React might get confused with the references when diffing and it might bring us unexpected behaviors.

I suggest creating a new variable and using the `backgroundColor` variable inside it.

p-jackson replied:
Nice catch!

zaguiini reviewed (commented):
Great work, looking fancy!

vykes-mac committed:
refactor code for better readability

vykes-mac committed:
convert image component from js to typescript

zaguiini commented on client/components/image/index.tsx:
```diff
@@ -0,0 +1,26 @@
+	event.currentTarget.style.display = 'none';
+};
+
+export default function Image( { alt, className, src, ...rest }: ImageProps ) {
```
You don't need to redefine the `img` tag props, you can use the ones that are exposed by React with `React.ComponentPropsWithoutRef<"img">`

```suggestion
type ImageProps = React.ComponentPropsWithoutRef< 'img' >

export default function Image( { alt, className, src, ...rest }: ImageProps ) {
```

zaguiini commented on client/components/image/index.tsx:
```diff
@@ -0,0 +1,26 @@
+	);
+}
+
+type ImageProps = {
```
This can be removed 👍

zaguiini commented on packages/components/src/site-thumbnail/use-dominant-color.tsx:
```diff
@@ -0,0 +1,28 @@
+		const fac = new FastAverageColor();
+
+		//get the dominant color of image from url
+		if ( url && url.length > 0 ) {
+			fac
+				.getColorAsync( url, {
+					ignoredColor: [
+						[ 255, 255, 255, 255 ], // white
+						[ 0, 0, 0, 255 ], // black
+					],
+					crossOrigin: 'anonymous',
+				} )
+				.then( ( color ) => setColor( color ) );
+		}
```
So, how expensive this operation is? I think we can memoize it since the `url` is a valid hash and a collision is impossible...

The idea is to create a `Map` outside this hook (but in the same file) that would be queried in the initial value of `useState` and inside this effect. If found, bail out of the hook body since the value is already computed. If not, proceed with execution and store the result in the map.

Does it make sense?

p-jackson commented on client/components/image/index.tsx:
```diff
@@ -0,0 +1,26 @@
+	event.currentTarget.style.display = 'none';
+};
+
+export default function Image( { alt, className, src, ...rest }: ImageProps ) {
```
I think we should still switch to this type which gives more correct typings since this component is designed to take all the same props as a `<img>` tag.

```suggestion
export default function Image( { alt, className, src, ...rest }: React.ImgHTMLAttributes< HTMLImageElement >  ) {
```

p-jackson replied:
Oops, I missed @zaguiini's [comment above](https://github.com/Automattic/wp-calypso/pull/66061#discussion_r939842724). His suggestion is better!

p-jackson commented on client/sites-dashboard/components/sites-table-row.tsx:
```diff
@@ -142,7 +147,23 @@ export default function SitesTableRow( { site }: SiteTableRowProps ) {
+									/>
+								) : (
+									<NoIcon role={ 'img' } aria-label={ __( 'Site Icon' ) }>
+										{ site.name.charAt( 0 ) }
```
I'm not sure this is 100% localisation friendly, but it'll probably be ok for now.

I think it'll work ok for words that start with simple diacritic mark. `site.name === Ā` for instance works fine.

This emoji for instance doesn't work (I guess I would have expected the icon to have been the complete emoji)
<img width="319" alt="CleanShot 2022-08-08 at 20 00 14@2x" src="https://user-images.githubusercontent.com/1500769/183369222-74d65216-3634-4e0b-b0b7-34b4c704e7c4.png">

Maybe emoji aren't the most important use case, but it might mean that there are non-latin languages where things don't work correctly. This example in Thai for instance:
<img width="254" alt="CleanShot 2022-08-08 at 19 58 48@2x" src="https://user-images.githubusercontent.com/1500769/183368990-e6d122e0-737a-466c-931f-358042880f90.png">
If you zoom in you can see that part of the first "character" (or grapheme) has been cut out of the icon. And I'm not sure what the significance of that will be.

Anyway, I guess we should just merge as is, but be aware this might need to change.

p-jackson replied:
Looks like the built-in `Intl` API might be able to help us
panfyZ-JL-p2_string-functions/#comment-1918

p-jackson replied:
The `Intl.Segmenter` type does it 👍 But it's not available on Firefox so we'd have to fallback to `charAt(0)` 😭

```typescript
function getFirstGrapheme( input: string ) {
	try {
		const segmenter = new Intl.Segmenter();
		const segments = segmenter.segment( input );
		return Array.from( segments )[0].segment;
	} catch {
		// Intl.Segmenter is not available in all browsers
		return input.charAt( 0 );
	}
}
```

vykes-mac committed:
memoize url colors to prevent extraction computation all the time

vykes-mac committed:
refactor to extract props from img tag

vykes-mac committed:
add function to handle special characters when striping the first letter of site name

zaguiini commented on packages/components/src/site-thumbnail/use-dominant-color.tsx:
```diff
@@ -1,14 +1,22 @@
+
 export default function useDominantColor( url?: string ) {
-	const [ color, setColor ] = useState< FastAverageColorResult >();
+	const [ color, setColor ] = useState< FastAverageColorResult | undefined >(
```
Let's use TS inference to avoid redundant typing

```suggestion
	const [ color, setColor ] = useState(
```

zaguiini commented on packages/components/src/site-thumbnail/use-dominant-color.tsx:
```diff
@@ -1,14 +1,22 @@
 export default function useDominantColor( url?: string ) {
-	const [ color, setColor ] = useState< FastAverageColorResult >();
+	const [ color, setColor ] = useState< FastAverageColorResult | undefined >(
+		( url && colorMap.get( url ) ) || undefined
```
This can be shortened to `url ? colorMap.get( url ) : undefined`.

Explanation: `get` returns `FastAverageColorResult | undefined` by default.

zaguiini commented on packages/components/src/site-thumbnail/use-dominant-color.tsx:
```diff
@@ -1,14 +1,22 @@
 		if ( url && url.length > 0 ) {
+			if ( colorMap.has( url ) ) {
+				return;
+			}
```
We can do this check earlier, even before the `FastAverageColor` instance is created.

vykes-mac committed:
refactor for better readability

zaguiini reviewed (approved):
Good to go 👍

danielbachhuber commented on client/sites-dashboard/components/sites-table-row.tsx:
```diff
@@ -142,7 +147,23 @@ export default function SitesTableRow( { site }: SiteTableRowProps ) {
-							<SiteIcon siteId={ site.ID } size={ 50 } />
+							<SiteThumbnail
+								mShotsUrl={ shouldUseScreenshot ? site.URL : undefined }
+								alt={ site.name }
```
For more semantic correctness, this `alt` probably should be a localized version of 'Screenshot of %s' or similar.

p-jackson replied:
You could be right, but I'm planning to put together a PR that makes sure the accessible labels for these links are working nicely. I'll read up what the best practices are, but I suspect that it's going to be more important to correctly label the link than to label the iamge.

danielbachhuber commented on client/components/image/index.tsx:
```diff
@@ -0,0 +1,20 @@
+import classnames from 'classnames';
+import { SyntheticEvent } from 'react';
+
+const hideImageOnError = ( event: SyntheticEvent< HTMLImageElement > ) => {
```
This seems to hide the site icon when the site icon loads fine:

<img width="1124" alt="image" src="https://user-images.githubusercontent.com/36432/183507967-705593ac-05fc-4995-a41c-9c2e170e53fc.png">

Is this intended?

danielbachhuber reviewed (commented):
🔥 So much fancy React in this PR! I only understand half of it 🙃

vykes-mac committed:
Merge branch 'trunk' into add/sites-thumbnails

p-jackson committed:
Silence Typescript errors to do with Intl.Segmenter

We usually target the es2020 type library (see
packages/calypso-typescript-config/ts-package.json) but we can opt in to
the es2022 library with a special comment. It's ok to opt-in to these
types because we're being careful to catch any runtime type errors.

However Intl.Segmenter was only added to the es2022 type library in
version 4.7 of Typescript. So until then we need to work around the type
error using `any` (or define our own types for Intl.Segmenter).
